realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -257,58 +261,66 @@ private void updateAsyncEmptyRealmObject() {
         }
     }
 
-    void notifyAllListeners() {
-        notifyTypeBasedListeners();
+    /**
+     * This method calls all registered listeners on Realm, RealmResults and RealmObjects.
+     *
+     * PREREQUISITE: Only call this method after all objects are up to date. This means:
+     * - `advance_read` was called on the Realm.
+     * - `RealmResults.syncIfNeeded()` was called when collecting RealmResults listeners.
+     *
+     * @param realmResultsToBeNotified list of all RealmResults listeners that can be notified.
+     */
+    void notifyAllListeners(List<RealmResults<? extends RealmModel>> realmResultsToBeNotified) {
 
-        // empty async RealmObject shouldn't block the realm to advance
-        // they're empty so no risk for running into a corrupt state
-        // where the pointer (Row) is using one version of a Realm, whereas the
-        // current Realm is advancing to a newer version (they're empty anyway)
+        // Notify all Realmresults (async and synchronous).
+        for (Iterator<RealmResults<? extends RealmModel>> it = realmResultsToBeNotified.iterator(); it.hasNext() && !realm.isClosed(); ) {
```

Comment:
Would it make sense to write `!realm.isClosed() && it.hasNext()` instead? I mean, is `realm` is closed, it sounds safer to terminate the loop.